In [2]:
import pandas as pd

In [5]:
data = pd.read_csv(r"C:\Users\ayush\OneDrive\Desktop\coding\100dy for ml\Day1\train.csv")

In [9]:
data.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Embarked  
0      0         A/5 21171   7.2500        S  
1      0          PC 17599  71.2833        C  
2      0  STON/O2. 3101282   7.9250        S  
3      0            113803  53.1000        S  
4      0            373450   8.0500        S

CHECK FOR MISSING VALUES

In [6]:
missing_values = data.isnull().sum()
print(missing_values)

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64


ONCE WE HAVE CHECKED THE NULL VALUES WE HAVE TO CHECK THE DATASET LENGTH

In [7]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


AS WE CAN SEE THERES Like 687 VALUES ARE MISSING IN CABIN AND ONLY 204 VALUES PRESENT IN CABIN , Simmilary IN AGEEMBARKEED

Imputation Techniques: When to use which one?
a. Dropping Rows or Columns with Missing Data:
When to use: If the feature has too many missing values and isn’t useful, or if missing data is a small proportion of your dataset.
In the Titanic dataset, the Cabin column has many missing values, making it hard to impute. It may be reasonable to drop it.


HENCE WE WILL DELETE ENTIRE COLUMN

In [8]:
data = data.drop(columns=['Cabin'])

AS WE CAN SEE ONLY 2 VALUE MISSING IN EMBARKED AND ITS CATEGORIAL DATA SO WE CAN REMOVE THE MISSING ROWS

In [12]:
data = data.dropna(subset=['Embarked'])

b. Replacing Missing Values with Statistical Measures (Mean, Median, Mode):
When to use:
Mean: Use for continuous numerical features if the distribution is approximately normal (e.g., Age).
Median: Use for continuous numerical features with skewed distributions (e.g., salary data, house prices).
Mode: Use for categorical features (e.g., Embarked).

In [13]:
data['Age'].fillna(data['Age'].mean(), inplace=True)

C:\Users\ayush\AppData\Local\Temp\ipykernel_25548\1981374165.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  data['Age'].fillna(data['Age'].mean(), inplace=True)


In [14]:
data['Age'].fillna(data['Age'].median(), inplace=True)

C:\Users\ayush\AppData\Local\Temp\ipykernel_25548\913469570.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  data['Age'].fillna(data['Age'].median(), inplace=True)


In [15]:
data['Embarked'].fillna(data['Embarked'].mode()[0], inplace=True)

C:\Users\ayush\AppData\Local\Temp\ipykernel_25548\4247733614.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  data['Embarked'].fillna(data['Embarked'].mode()[0], inplace=True)


NOW LETS AGAIN SEE THE DATASET

In [18]:
data.isnull().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Embarked       0
dtype: int64

AS YOU CAN SEE WE HANDLED THE MISIING VALUE WITH IMPUTATION

THERE ARE SOME OTHER ADVANCED METHOD ALSO FOR FILLING MISIING VALUES

Advanced Methods (Forward/Backward Fill or K-Nearest Neighbors Imputation):
When to use:
Forward/Backward Fill: Use for time series or ordered data where the next or previous value is a reasonable estimate.
K-Nearest Neighbors (KNN): Use when the relationships between features can help predict the missing values. This method looks at the nearest data points to impute the missing value.

Forward Fill (ffill)
Forward fill propagates the last valid observation forward to fill gaps. For example, if you have a missing value in a column, it will be replaced with the last non-missing value that appeared before it.

Backward Fill (bfill)
Backward fill, on the other hand, propagates the next valid observation backward to fill gaps. This means that a missing value will be replaced with the next non-missing value that appears after it.

LETS RELOAD OUR DATASET AND RERUN WITH ANOTHER METHOD

In [19]:
data = pd.read_csv(r"C:\Users\ayush\OneDrive\Desktop\coding\100dy for ml\Day1\train.csv")

In [20]:
data.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [21]:
data.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [23]:
# Forward fill
data['Age'].fillna(method='ffill', inplace=True)
# Backward fill
data['Age'].fillna(method='bfill', inplace=True)

C:\Users\ayush\AppData\Local\Temp\ipykernel_25548\784571897.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  data['Age'].fillna(method='ffill', inplace=True)
C:\Users\ayush\AppData\Local\Temp\ipykernel_25548\784571897.py:2: FutureWarning: Series.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  data['Age'].fillna(method='ffill', inplace=True)
C:\Users\ayush\AppData\Local\Temp\ipykernel_25548\784571897.py:5: FutureWarning: A value is trying to be set on a copy of

In [24]:
data.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age              0
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

KNN IMPUTATION

In [26]:
from sklearn.impute import KNNImputer

# Select features to use for KNN imputation
features = data[['Pclass', 'SibSp', 'Parch', 'Fare', 'Age']]

# Apply KNN Imputer
imputer = KNNImputer(n_neighbors=3)
df_knn_imputed = imputer.fit_transform(features)

# Update the Age column with imputed values
data['Age'] = df_knn_imputed[:, 4]  # Column 4 corresponds to 'Age'


When to Use Each Method:
Drop rows/columns: When data is sparse or the column/row isn't crucial for analysis (e.g., Cabin).
Mean/Median imputation: For numerical columns with relatively low missing values and predictable distribution (e.g., Age).
Mode imputation: For categorical columns (e.g., Embarked).
Forward/Backward fill: For ordered data, such as time series.
KNN imputation: For more sophisticated imputation when relationships bet
ween features are important and can provide meaningful estimates (e.g., Age based on other personal characteristics).

OTHER METHODS THAT USUALLY USE FOR FILLINS MISSING VALUES ARE


Advanced Techniques:

    Multiple Imputation: Create multiple complete datasets by imputing missing values several times and then combine the results. This accounts for the uncertainty in the imputations.

    Model-Based Methods: Use algorithms that can handle missing data internally, such as decision trees or certain ensemble methods.

Encoding Missingness:

    Missing Indicator: Create a new feature that indicates whether a value was missing. This can help the model learn patterns related to missing data.

Predictive Modeling:

    Regression Imputation: Use regression models to predict and fill in missing values based on other available data1